src/ott/solvers/nn/potential_base.py

```diff
@@ -0,0 +1,113 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Base module for potentials."""
+
+import abc
+from typing import Callable, Optional
+
+import flax.linen as nn
+import jax
+import jax.numpy as jnp
+from flax import struct
+from flax.core import frozen_dict
+from flax.training import train_state
+
+__all__ = ["PotentialBase", "PotentialTrainState"]
+
+
+class PotentialTrainState(train_state.TrainState):
+  """Adds information about the potential value and gradient to the state."""
+  potential_value_fn: Callable[[jnp.ndarray],
+                               jnp.ndarray] = struct.field(pytree_node=False)
+  potential_gradient_fn: Callable[[jnp.ndarray], jnp.ndarray] = struct.field(
+      pytree_node=False
+  )
+
+
+class PotentialBase(abc.ABC, nn.Module):
+  """Base class for potentials."""
+
+  @property
+  @abc.abstractmethod
+  def is_potential(self) -> bool:
+    """Indicates if the module defines the potential's value or the gradient.
+
+    Returns:
+      ``True`` if the module defines the potential's value, ``False``
+      if it defined the gradient.
+    """
+
+  def potential_value(
```

Comment:
I wonder whether this should also evaluate it on `x` or just return the potential function, as I think it will be more jax-y; would return `Callable[[jnp.ndarray], jnp.ndarray]`.

Reply:
I made it behave similar to the `apply` function, which also evaluates the point. This way they're stored along in the training state in the same way as the `apply` function rather than needing an additional call to create the function. Do you think that's reasonable?

Reply:
Actually I thought about it some more and it seems a little nicer to make them functions. I just pushed this in https://github.com/ott-jax/ott/pull/219/commits/ce90b67ff4f755ad664bf4deede78584717160fb and will keep this thread open in case you have any other thoughts here